lib/block_view.go

```diff
@@ -6930,6 +7148,314 @@ func (bav *UtxoView) _connectNFTBid(
+		return 0, 0, nil, fmt.Errorf("_connectNFTTransfer: non-existent updaterPKID: %s",
+			PkToString(txn.PublicKey, bav.Params))
+	}
+	if !reflect.DeepEqual(prevNFTEntry.OwnerPKID, updaterPKID.PKID) {
+		return 0, 0, nil, RuleErrorNFTTransferByNonOwner
+	}
+
+	// Fetch the receiver's PKID and make sure it exists.
+	receiverPKID := bav.GetPKIDForPublicKey(txMeta.ReceiverPublicKey)
+	// Sanity check that we found a PKID entry for these pub keys (should never fail).
+	if receiverPKID == nil || receiverPKID.isDeleted {
+		return 0, 0, nil, fmt.Errorf(
+			"_connectNFTTransfer: Found nil or deleted PKID for receiver, this should never "+
+				"happen. Receiver pubkey: %v", PkToStringMainnet(txMeta.ReceiverPublicKey))
+	}
+
+	// Make sure that the NFT entry is not for sale.
+	if prevNFTEntry.IsForSale {
+		return 0, 0, nil, RuleErrorCannotTransferForSaleNFT
+	}
+
+	// Sanity check that the NFT entry is correct.
+	if !reflect.DeepEqual(prevNFTEntry.NFTPostHash, txMeta.NFTPostHash) ||
+		!reflect.DeepEqual(prevNFTEntry.SerialNumber, txMeta.SerialNumber) {
+		return 0, 0, nil, fmt.Errorf("_connectNFTTransfer: prevNFTEntry %v is inconsistent with txMeta %v;"+
+			" this should never happen.", prevNFTEntry, txMeta)
+	}
+
+	// Get the postEntry so we can check for unlockable content.
+	nftPostEntry := bav.GetPostEntryForPostHash(txMeta.NFTPostHash)
+	if nftPostEntry == nil || nftPostEntry.isDeleted {
+		return 0, 0, nil, fmt.Errorf("_connectNFTTransfer: non-existent nftPostEntry for NFTPostHash: %s",
+			txMeta.NFTPostHash.String())
+	}
+
+	// If the post entry requires the NFT to have unlockable text, make sure it is provided.
+	if nftPostEntry.HasUnlockable && len(txMeta.UnlockableText) == 0 {
+		return 0, 0, nil, RuleErrorCannotTransferUnlockableNFTWithoutUnlockable
+	}
+
+	// Check the length of the UnlockableText.
+	if uint64(len(txMeta.UnlockableText)) > bav.Params.MaxPrivateMessageLengthBytes {
+		return 0, 0, nil, errors.Wrapf(
+			RuleErrorUnlockableTextLengthExceedsMax, "_connectNFTTransfer: "+
+				"UnlockableTextLen = %d; Max length = %d",
+			len(txMeta.UnlockableText), bav.Params.MaxPrivateMessageLengthBytes)
+	}
+
+	// Connect basic txn to get the total input and the total output without
+	// considering the transaction metadata.
+	totalInput, totalOutput, utxoOpsForTxn, err := bav._connectBasicTransfer(
+		txn, txHash, blockHeight, verifySignatures)
+	if err != nil {
+		return 0, 0, nil, errors.Wrapf(err, "_connectNFTTransfer: ")
+	}
+
+	// Force the input to be non-zero so that we can prevent replay attacks.
+	if totalInput == 0 {
+		return 0, 0, nil, RuleErrorNFTTransferRequiresNonZeroInput
+	}
+
+	if verifySignatures {
+		// _connectBasicTransfer has already checked that the transaction is
+		// signed by the top-level public key, which we take to be the poster's
+		// public key.
+	}
+
+	// Now we are ready to transfer the NFT.
+
+	// Create the updated NFTEntry.
+	newNFTEntry := &NFTEntry{
```

Comment:
Hmm... this seems error-prone compared to setting only the fields that are changing. For example, if we add a new field to NFTEntry then we could forget to initialize it here. Is there a reason you did it this way vs the following?

```
// Make a copy of the previous NFT
newNFTEntry := *prevNFTEntry
// Update the fields that were set during this transfer.
newNFTEntry.OwnerPKID = receiverPKID.PKID
newNFTEntry.UnlockableText = txMeta.UnlockableTex
newNFTEntry.IsPencing = true
```

Reply:
No reason, just copied it from _connectCreateNFT.  Changed.